sanitizers: Stabilize AddressSanitizer and LeakSanitizer (#123617)
rust-lang/rust: src/doc/rustc/src/codegen-options/index.md

* `y`, `yes`, `on`, `true` or no value: enable rpath.
* `n`, `no`, `off` or `false`: disable rpath (the default).

## sanitize

Sanitizers are a set of compiler-based runtime error detection tools that
instrument programs to detect bugs during execution. They work by instrumenting
code at compile time and runtime to monitor program behavior and detect specific
classes of errors at runtime. Sanitizers enable precise, low-overhead runtime
bug detection, improving software reliability and security.

This option allows for use of one or more of these sanitizers:

  * [AddressSanitizer

>     (ASan)](https://doc.rust-lang.org/rustc/sanitizers.html#addresssanitizer):

Reviewer: All links to books should be relative (this helps with offline viewing and linkchecking).

```suggestion
    (ASan)](../sanitizers.md#addresssanitizer):
```

    Detects memory errors (e.g., buffer overflows, use after free).
  * [LeakSanitizer

>     (LSan)](https://doc.rust-lang.org/rustc/sanitizers.html#leaksanitizer):

Reviewer: ```suggestion
    (LSan)](../sanitizers.md#leaksanitizer):
```

    Detects memory leaks either as part of AddressSanitizer or as a standalone
    tool.

These are the valid values for this option for targets that support one or more
of these sanitizers:

| Target                      | Sanitizers      |
|-----------------------------|-----------------|
| aarch64-apple-darwin        | address         |
| aarch64-unknown-linux-gnu   | address, leak   |
| i686-pc-windows-msvc        | address         |
| i686-unknown-linux-gnu      | address         |
| x86_64-apple-darwin         | address, leak   |
| x86_64-pc-windows-msvc      | address         |
| x86_64-unknown-linux-gnu    | address, leak   |

The quality of the Sanitizers implementation and support varies across operating
systems and architectures, and relies heavily on LLVM implementation--they are
mostly implemented in and supported by LLVM.

Using a different LLVM or runtime version than the one used by the Rust compiler
is not supported. Using Sanitizers in mixed-language binaries (also known as
“mixed binaries”) is supported when the same LLVM and runtime version is used by
all languages.

> For more information about the Sanitizers, see the [Sanitizers
> chapter](https://doc.rust-lang.org/rustc/sanitizers.html) of [The rustc
> book](https://doc.rust-lang.org/rustc/).

Reviewer: ```suggestion
For more information about the Sanitizers, see the [Sanitizers
chapter](../sanitizers.md).
```

Reviewer 2: ```suggestion
For more information about sanitizers, see the [Sanitizers
chapter](../sanitizers.md).
```

## sanitize

Sanitizers are tools that help detect and prevent various types of bugs and
vulnerabilities in software. They are available in compilers and work by
instrumenting the code to add additional runtime checks. While they provide
powerful tools for identifying bugs or security issues, it's important to note
that using sanitizers can introduce runtime overhead and might not catch all
possible issues. Therefore, they are typically used alongside other best
practices in software development, such as testing and fuzzing, to ensure the
highest level of software quality and security.

This option allows for use of one or more of these sanitizers:

* [AddressSanitizer](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/sanitizer.html#addresssanitizer)
  a fast memory error detector.
* [LeakSanitizer](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/sanitizer.html#leaksanitizer)
  a run-time memory leak detector.

These are the valid values for this option for targets that support one or more
of these sanitizers:

| Target | Sanitizers |
|--------|------------|
| aarch64-unknown-linux-gnu | address, leak |

> | i586-pc-windows-msvc | address |

Reviewer 3: `i586-pc-windows-msvc` isn't changed in this PR or a Tier 1 target

```suggestion
```

Author: Oops, fixed.

> | i586-unknown-linux-gnu | address |

Reviewer 3: Right?

```suggestion
| i686-unknown-linux-gnu | address |
```

Author: Done.

| i686-pc-windows-msvc | address |
| i686-unknown-linux-gnu | address |
| x86_64-apple-darwin | address, leak |
| x86_64-pc-windows-msvc | address |
| x86_64-unknown-linux-gnu | address, leak |

For more information on sanitizers, see the
[sanitizer
feature](https://doc.rust-lang.org/nightly/unstable-book/compiler-flags/sanitizer.html)
in the [The Rust Unstable

> Book](https://doc.rust-lang.org/nightly/unstable-book/). For more information on

Reviewer 3: I think we should move the relevant documentation for stable sanitizers into this document (we should not need to refer people to the unstable book for information on stable features 🙂) 

In particular, the section ASan should be moved and I think it would be extremely helpful to have a section on some of the considerations that need to be taken into account to use sanitizers in mixed C/C++/Rust codebases as well as a general disclaimer that the quality of the sanitizer support varies across platforms and relies heavily on LLVM's implementation.

Author: Sorry for the delay again. Done. (And as we've discussed, I've also started writing a new Sanitizers chapter for The rustc book.)

stable and supported sanitizers, see the [supported sanitizers

> table](https://github.com/rust-lang/rust/issues/123615#issuecomment-2041791236).

Reviewer 3: This table should live in this doc as well.

Reviewer 3: Oh, this table already exists above? I think this sentence can just be removed then?

Author: Done.

* `y`, `yes`, `on`, `true` or no value: enable rpath.
* `n`, `no`, `off` or `false`: disable rpath (the default).

> ## sanitizer

Author: Done.
